src/units/abilities.cpp

```diff
@@ -402,39 +408,68 @@ bool unit::has_ability_type(const std::string& ability) const
 	return !abilities_.child_range(ability).empty();
 }
 
-
-std::pair<int,map_location> unit_ability_list::highest(const std::string& key, int def) const
+namespace {
+	
+	
+template<typename T, typename TFuncFormula>
+class get_ability_value_visitor : public boost::static_visitor<T>
 {
-	if ( cfgs_.empty() ) {
-		return std::make_pair(def, map_location());
-	}
-	// The returned location is the best non-cumulative one, if any,
-	// the best absolute cumulative one otherwise.
-	map_location best_loc;
-	bool only_cumulative = true;
-	int abs_max = 0;
-	int flat = 0;
-	int stack = 0;
-	for (unit_ability const &p : cfgs_)
+public:
+	// Constructor stores the default value.
+	get_ability_value_visitor(T def, const TFuncFormula& formula_handler) : def_(def), formula_handler_(formula_handler) {}
+
+	T operator()(boost::blank const &) const { return def_; }
+	T operator()(bool)                 const { return def_; }
+	T operator()(int i)                const { return static_cast<T>(i); }
+	T operator()(unsigned long long u) const { return static_cast<T>(u); }
+	T operator()(double d)             const { return static_cast<T>(d); }
+	T operator()(t_string const &)     const { return def_; }
+	T operator()(const std::string& s) const
 	{
-		int value = (*p.first)[key].to_int(def);
-		if ((*p.first)["cumulative"].to_bool()) {
-			stack += value;
-			if (value < 0) value = -value;
-			if (only_cumulative && value >= abs_max) {
-				abs_max = value;
-				best_loc = p.second;
-			}
-		} else if (only_cumulative || value > flat) {
-			only_cumulative = false;
-			flat = value;
-			best_loc = p.second;
+		utils::string_view formula_prefix("formula:");
```

Comment:
I don't think the first complaint really matters - if you really think it's a problem, it can be fixed in  `typed_formula`, anyway.

As for the second complaint... no other place provides WFL and Lua support with the same syntax, so I don't see any particular reason why this one should be any different. Why not use a `lua_function` key for Lua support? Or even a `[lua]` tag?